compose_actions: Hide compose box when navigating to inbox/recent view.

sayamsamal opened the pull request:
Currently, given that the compose box is already open, it stays open when the user navigates to the inbox/recent view, even when there are no modifications to the message or recipients.

This PR adds conditions to close the compose box when navigating to the inbox/recent view, except when we can reasonably infer that it was the intended action.

<!-- Describe your pull request here.-->

Fixes: [Issue reported at CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/compose.20box.20tooltips.20lag.20.2327158/near/1697754)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

## Navigating between recent/inbox view with different compose states
### **Before**

[before-compose-actions-fix.webm](https://github.com/zulip/zulip/assets/82862779/dddcc09a-6894-43e6-a0cd-bcf1d9fd7a4c)

### **After**

[after-compose-actions-fix.webm](https://github.com/zulip/zulip/assets/82862779/3b6881de-1c22-417c-932e-eaf50cd14fee) 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/hashchange.js:
```diff
@@ -50,6 +51,18 @@ function maybe_hide_inbox() {
+    compose_actions.on_view_change();
+    maybe_hide_inbox();
+    recent_view_ui.show();
+}
```
We already have a `views_util` module intended for this sort of common code to be called when switching into inbox or recent conversations; let's use that rather than the refactoring in this file.

sayamsamal replied:
Thank you for the insight, that also helped simplify the code. Addressed in the new code.

sayamsamal replied:
Should we also add a comment line for the new `on_view_change` function? I've omitted it atm because I felt that the logic could be inferred from the adjacent functions - `on_topic_narrow` and `on_narrow`.

timabbott replied:
I think the new function could possibly benefit from a comment, also posted a comment about naming; otherwise lgtm.

sayamsamal commented:
Updated screencast (visually same as before; intended to show that the new code still works)

[screencast-compose-actions-update-1.webm](https://github.com/zulip/zulip/assets/82862779/0cb1b98c-286e-4517-bae0-e60ff0e72ac8)

timabbott commented on web/src/compose_actions.js:
```diff
@@ -336,6 +336,16 @@ export function cancel() {
     $(document).trigger("compose_canceled.zulip");
 }
 
+export function on_view_change() {
```
Let's call this `on_show_navigation_view` -- I don't want any ambiguity about whether a narrow qualifies.

timabbott commented:
Managed to trigger an exception clicking around randomly:

![image](https://github.com/zulip/zulip/assets/2746074/e204d008-7a9a-45d0-8371-80d5b94de5cd)

I think what's going on is that this is being done too early; the `compose_canceled.zulip` hook expects the recent view to be either fully not the current state or fully the current state, so we need to do this after the `complete_rerender` hook.

timabbott commented:
Pushed my changes and test-deploying on chat.zulip.org.

sayamsamal commented:
Thank you for catching that bug; I was also able to reproduce the same a few moments ago. I have added appropriate comments to the new function and updated the code slightly to match the logic in the surrounding functions. (change conditional to guard clause)

timabbott commented:
Merged, thanks @sayamsamal!
